nodist/clang_check_attributes.py

```diff
@@ -0,0 +1,325 @@
+    return True
+
+
+def print_funcs(node, cfg, search_attributes, level=1):
+    funcs = get_call_expr(node)
+    for func in funcs:
+        if not func.referenced:
+            continue
+        if is_excluded(func.referenced, cfg):
+            continue
+        call = func.referenced
+        # Don't recurse into already printed nodes
+        # it some cases in increases printing nodes by a lot
+        if fully_qualified_pretty(call) in PRINTED:
+            continue
+        if str(call.location.file).endswith(".h"):
+            if os.path.exists(str(call.location.file).replace(".h", ".cpp")):
+                tu_call = find_usr(
+                    str(call.location.file).replace(".h", ".cpp"),
+                    call.get_usr(), cfg)
+                if tu_call is not None:
+                    call = tu_call
+
+        if call is not None:
+            print_node(call, level)
+            if call.kind in [
+                    CursorKind.CXX_METHOD, CursorKind.CONSTRUCTOR,
+                    CursorKind.FUNCTION_DECL
+            ]:
+                print_funcs(call, cfg, search_attributes, level + 1)
+
+
+def get_info(node, cfg):
+    if node.kind == CursorKind.CXX_METHOD:
+        annotations = get_annotations(node)
+        if "MT_SAFE" in annotations:
+            if not find_funcs(node, cfg, ["MT_SAFE", "PURE"]):
+                UNSAFE_CALLS.append(node)
+        if "PURE" in annotations:
+            if not find_funcs(node, cfg, ["PURE"]):
+                UNSAFE_CALLS.append(node)
+        if "MT_START" in annotations:
+            if not find_funcs(node, cfg, ["MT_SAFE", "PURE"]):
+                UNSAFE_CALLS.append(node)
+    for cursor in node.get_children():
+        get_info(cursor, cfg)
+
+
+def get_files(filename):
+    files = list(glob.glob(f"{filename}/*.cpp"))
+    # Dont process test files
+    files = [f for f in files if "_test.cpp" not in f]
+    # Dont process Verilator coverage files
+    files = [f for f in files if "Vlc" not in f]
+    # We want to use V3Const after astgen
+    files = [f for f in files if "V3Const.cpp" not in f]
+    # Skip SystemC files
+    files = [f for f in files if "_sc" not in f]
+    # Skip parser files
+    files = [f for f in files if ".yy.cpp" not in f]
+    # Make sure list is unique
+    files = list(dict.fromkeys(files))
+    # Make filepaths absolute
+    files = [os.path.abspath(f) for f in files]
+    return files
+
+
+def read_args(args):
+    base_folders = []
+    clang_args = []
+    excluded_prefixes = []
+    excluded_paths = ['/usr']
+    i = 0
+    while i < len(args):
+        if args[i] == '-x':
+            i += 1
+            excluded_prefixes = args[i].split(',')
+        elif args[i] == '-p':
+            i += 1
+            excluded_paths = args[i].split(',')
+        elif args[i][0] == '-':
+            clang_args.append(args[i])
+        else:
+            base_folders.append(args[i])
+        i += 1
```

Comment:
Please use https://docs.python.org/3/library/argparse.html
for examples see the other python programs in Verilator, e.g. src/arggen.